File: assets/js/lib/delta.js
/**
 * Delta is a format used to represent a set of changes introduced to a text document.
 *
 * See `Livebook.Delta` for more details.
 *
 * An implementation of the full Delta specification is available
 * in the official quill-delta package (https://github.com/quilljs/delta)
 * licensed under MIT. Our version is based on that package,
 * but simplified to match our non rich-text use case.
 */
export default class Delta {
  constructor(ops = []) {
    this.ops = ops;
  }

  /**
   * Appends a retain operation.
   */
  retain(length) {
    if (length <= 0) {
      return this;
    }

    return this.append({ retain: length });
  }

  /**
   * Appends an insert operation.
   */
  insert(text) {
    if (text === "") {
      return this;
    }

    return this.append({ insert: text });
  }

  /**
   * Appends a delete operation.
   */
  delete(length) {
    if (length <= 0) {
      return this;
    }

    return this.append({ delete: length });
  }

  /**
   * Appends the given operation.
   *
   * See `Livebook.Delta.append/2` for more details.
   */
  append(op) {
    if (this.ops.length === 0) {
      this.ops.push(op);
      return this;
    }

    const lastOp = this.ops.pop();

    // Insert and delete are commutative, so we always make sure
    // to put insert first to preserve the canonical form.
    if (isInsert(op) && isDelete(lastOp)) {
      return this.append(op).append(lastOp);
    }

    if (isInsert(op) && isInsert(lastOp)) {
      this.ops.push({ insert: lastOp.insert + op.insert });
      return this;
    }

    if (isDelete(op) && isDelete(lastOp)) {
      this.ops.push({ delete: lastOp.delete + op.delete });
      return this;
    }

    if (isRetain(op) && isRetain(lastOp)) {
      this.ops.push({ retain: lastOp.retain + op.retain });
      return this;
    }

    this.ops.push(lastOp, op);
    return this;
  }

  /**
   * Returns a new delta that is equivalent to applying the operations of this delta,
   * followed by operations of the given delta.
   */
  compose(other) {
    const thisIter = new Iterator(this.ops);
    const otherIter = new Iterator(other.ops);
    const delta = new Delta();

    while (thisIter.hasNext() || otherIter.hasNext()) {
      if (isInsert(otherIter.peek())) {
        delta.append(otherIter.next());
      } else if (isDelete(thisIter.peek())) {
        delta.append(thisIter.next());
      } else {
        const length = Math.min(thisIter.peekLength(), otherIter.peekLength());
        const thisOp = thisIter.next(length);
        const otherOp = otherIter.next(length);

        if (isRetain(otherOp)) {
          // Either retain or insert, so just apply it.
          delta.append(thisOp);

          // Other op should be delete, we could be an insert or retain
          // Insert + delete cancels out
        } else if (isDelete(otherOp) && isRetain(thisOp)) {
          delta.append(otherOp);
        }
      }
    }

    return delta._trim();
  }

  /**
   * Transform the given delta against this delta's operations. Returns a new delta.
   *
   * The method takes a `priority` argument indicates which delta
   * is considered to have happened first and is used for conflict resolution.
   *
   * See `Livebook.Delta.Transformation` for more details.
   */
  transform(other, priority) {
    if (priority !== "left" && priority !== "right") {
      throw new Error(
        `Invalid priority "${priority}", should be either "left" or "right"`
      );
    }

    const thisIter = new Iterator(this.ops);
    const otherIter = new Iterator(other.ops);
    const delta = new Delta();

    while (thisIter.hasNext() || otherIter.hasNext()) {
      if (
        isInsert(thisIter.peek()) &&
        (!isInsert(otherIter.peek()) || priority === "left")
      ) {
        const insertLength = operationLength(thisIter.next());
        delta.retain(insertLength);
      } else if (isInsert(otherIter.peek())) {
        delta.append(otherIter.next());
      } else {
        const length = Math.min(thisIter.peekLength(), otherIter.peekLength());
        const thisOp = thisIter.next(length);
        const otherOp = otherIter.next(length);

        if (isDelete(thisOp)) {
          // Our delete either makes their delete redundant or removes their retain
          continue;
        } else if (isDelete(otherOp)) {
          delta.append(otherOp);
        } else {
          // We retain either their retain or insert
          delta.retain(length);
        }
      }
    }

    return delta._trim();
  }

  _trim() {
    if (this.ops.length > 0 && isRetain(this.ops[this.ops.length - 1])) {
      this.ops.pop();
    }

    return this;
  }

  /**
   * Converts the delta to a compact representation, suitable for sending over the network.
   */
  toCompressed() {
    return this.ops.map((op) => {
      if (isInsert(op)) {
        return op.insert;
      } else if (isRetain(op)) {
        return op.retain;
      } else if (isDelete(op)) {
        return -op.delete;
      }

      throw new Error(`Invalid operation ${op}`);
    });
  }

  /**
   * Builds a new delta from the given compact representation.
   */
  static fromCompressed(list) {
    return list.reduce((delta, compressedOp) => {
      if (typeof compressedOp === "string") {
        return delta.insert(compressedOp);
      } else if (typeof compressedOp === "number" && compressedOp >= 0) {
        return delta.retain(compressedOp);
      } else if (typeof compressedOp === "number" && compressedOp < 0) {
        return delta.delete(-compressedOp);
      }

      throw new Error(`Invalid compressed operation ${compressedOp}`);
    }, new this());
  }

  /**
   * Returns the result of applying the delta to the given string.
   */
  applyToString(string) {
    let newString = "";
    let index = 0;

    this.ops.forEach((op) => {
      if (isRetain(op)) {
        newString += string.slice(index, index + op.retain);
        index += op.retain;
      }

      if (isInsert(op)) {
        newString += op.insert;
      }

      if (isDelete(op)) {
        index += op.delete;
      }
    });

    newString += string.slice(index);

    return newString;
  }
}

/**
 * Operations iterator simplifying the implementation of the delta methods above.
 *
 * Allows for iterating over operation slices by specifying the desired length.
 */
class Iterator {
  constructor(ops) {
    this.ops = ops;
    this.index = 0;
    this.offset = 0;
  }

  hasNext() {
    return this.peekLength() < Infinity;
  }

  next(length = Infinity) {
    const nextOp = this.ops[this.index];

    if (nextOp) {
      const offset = this.offset;
      const opLength = operationLength(nextOp);

      if (length >= opLength - offset) {
        length = opLength - offset;
        this.index += 1;
        this.offset = 0;
      } else {
        this.offset += length;
      }

      if (isDelete(nextOp)) {
        return { delete: length };
      } else if (isRetain(nextOp)) {
        return { retain: length };
      } else if (isInsert(nextOp)) {
        return { insert: nextOp.insert.substr(offset, length) };
      }
    } else {
      return { retain: length };
    }
  }

  peek() {
    return this.ops[this.index] || { retain: Infinity };
  }

  peekLength() {
    if (this.ops[this.index]) {
      return operationLength(this.ops[this.index]) - this.offset;
    } else {
      return Infinity;
    }
  }
}

function operationLength(op) {
  if (isInsert(op)) {
    return op.insert.length;
  }

  if (isRetain(op)) {
    return op.retain;
  }

  if (isDelete(op)) {
    return op.delete;
  }
}

export function isInsert(op) {
  return typeof op.insert === "string";
}

export function isRetain(op) {
  return typeof op.retain === "number";
}

export function isDelete(op) {
  return typeof op.delete === "number";
}

File: lib/livebook/text/delta.ex
defmodule Livebook.Text.Delta do
  # Delta is a format used to represent a set of changes introduced
  # to a text document.
  #
  # By design, delta is suitable for Operational Transformation and
  # is hence our primary building block in collaborative text editing.
  #
  # For a detailed write-up see https://quilljs.com/docs/delta and
  # https://quilljs.com/guides/designing-the-delta-format. The
  # specification covers rich-text editing, while we only need to
  # work with plain-text, so we use a subset of the specification
  # with operations listed in `Livebook.Text.Delta.Operation`.
  #
  # An implementation of the full Delta specification is available in
  # the :text_delta package (https://github.com/deltadoc/text_delta)
  # by Konstantin Kudryashov under the MIT license. This module builds
  # directly on that package, and is simplified to better fit our not
  # rich-text use case.
  #
  # ## Lengths and offsets
  #
  # Delta operations involve strings and string lengths. The exact
  # definition of these depends on the string representation in the
  # given programming language. This module is implemented for
  # compatibility with JavaScript, hence all lengths match the length
  # as defined by JavaScript (the number of UTF-16 code units).
  #
  # All places with calls to the `Livebook.Text.JS` module are the
  # ones where the implementation is affected by this distinction.

  defstruct ops: []

  alias Livebook.Text
  alias Livebook.Text.Delta
  alias Livebook.Text.Delta.Operation
  alias Livebook.Text.Delta.Transformation

  @typedoc """
  Delta carries a list of consecutive operations.

  Note that we keep the operations in reversed order for efficiency.
  """
  @type t :: %Delta{ops: list(Operation.t())}

  @doc """
  Creates a new delta, optionally taking a list of operations.
  """
  @spec new(list(Operation.t())) :: t()
  def new(opts \\ [])
  def new([]), do: %Delta{}
  def new(ops), do: Enum.reduce(ops, new(), &append(&2, &1))

  @doc """
  Appends a new `:insert` operation to the given delta.
  """
  @spec insert(t(), String.t()) :: t()
  def insert(delta, string) do
    append(delta, Operation.insert(string))
  end

  @doc """
  Appends a new `:retain` operation to the given delta.
  """
  @spec retain(t(), non_neg_integer()) :: t()
  def retain(delta, length) do
    append(delta, Operation.retain(length))
  end

  @doc """
  Appends a new `:delete` operation to the given delta.
  """
  @spec delete(t(), non_neg_integer()) :: t()
  def delete(delta, length) do
    append(delta, Operation.delete(length))
  end

  @doc """
  Appends an operation to the given delta.

  The specification imposes two constraints:

    1. Delta must be *compact* - there must be no shorter equivalent
       delta.

    2. Delta must be *canonical* - there is just a single valid
       representation of the given change.

  To satisfy these constraints we follow two rules:

    1. Delete followed by insert is swapped to ensure that insert
       goes first.

    2. Operations of the same type are merged.

  """
  @spec append(t(), Operation.t()) :: t()
  def append(delta, op) do
    Map.update!(delta, :ops, &compact(&1, op))
  end

  defp compact(ops, {:insert, ""}), do: ops
  defp compact(ops, {:retain, 0}), do: ops
  defp compact(ops, {:delete, 0}), do: ops
  defp compact([], new_op), do: [new_op]

  defp compact([{:delete, _} = del | ops_remainder], {:insert, _} = ins) do
    ops_remainder
    |> compact(ins)
    |> compact(del)
  end

  defp compact([{:retain, len_a} | ops_remainder], {:retain, len_b}) do
    [Operation.retain(len_a + len_b) | ops_remainder]
  end

  defp compact([{:insert, str_a} | ops_remainder], {:insert, str_b}) do
    [Operation.insert(str_a <> str_b) | ops_remainder]
  end

  defp compact([{:delete, len_a} | ops_remainder], {:delete, len_b}) do
    [Operation.delete(len_a + len_b) | ops_remainder]
  end

  defp compact(ops, new_op), do: [new_op | ops]

  @doc """
  Removes trailing retain operations from the given delta.
  """
  @spec trim(t()) :: t()
  def trim(%Delta{ops: [{:retain, _} | ops]} = delta), do: %{delta | ops: ops}
  def trim(delta), do: delta

  @doc """
  Checks if the delta has no changes.
  """
  @spec empty?(t()) :: boolean()
  def empty?(delta) do
    trim(delta).ops == []
  end

  @doc """
  Returns data operations in the order in which they apply.
  """
  @spec operations(t()) :: list(Operation.t())
  def operations(delta) do
    Enum.reverse(delta.ops)
  end

  @doc """
  Converts the given delta to a compact representation, suitable for
  JSON serialization.

  ## Examples

      iex> delta = Delta.new([retain: 2, insert: "hey", delete: 3])
      iex> Livebook.Text.Delta.to_compressed(delta)
      [2, "hey", -3]

  """
  @spec to_compressed(t()) :: list(Operation.compressed_t())
  def to_compressed(delta) do
    delta.ops
    |> Enum.reverse()
    |> Enum.map(&Operation.to_compressed/1)
  end

  @doc """
  Builds a new delta from the given compact representation.

  ## Examples

      iex> delta = Livebook.Text.Delta.from_compressed([2, "hey", -3])
      iex> Livebook.Text.Delta.operations(delta)
      [retain: 2, insert: "hey", delete: 3]

  """
  @spec from_compressed(list(Operation.compressed_t())) :: t()
  def from_compressed(list) do
    ops =
      list
      |> Enum.map(&Operation.from_compressed/1)
      |> Enum.reverse()

    %Delta{ops: ops}
  end

  defdelegate transform(left, right, priority), to: Transformation

  defdelegate transform_position(delta, index), to: Transformation

  @doc """
  Returns the result of applying `delta` to `string`.
  """
  @spec apply(Delta.t(), String.t()) :: String.t()
  def apply(delta, string) do
    do_apply(operations(delta), <<>>, string)
  end

  defp do_apply([{:retain, n} | ops], result, string) do
    {left, right} = Text.JS.split_at(string, n)
    do_apply(ops, <<result::binary, left::binary>>, right)
  end

  defp do_apply([{:insert, inserted} | ops], result, string) do
    do_apply(ops, <<result::binary, inserted::binary>>, string)
  end

  defp do_apply([{:delete, n} | ops], result, string) do
    do_apply(ops, result, Text.JS.slice(string, n))
  end

  defp do_apply([], result, string) do
    <<result::binary, string::binary>>
  end

  @doc """
  Computes Myers Difference between the given strings and returns its
  `Delta` representation.
  """
  @spec diff(String.t(), String.t()) :: Delta.t()
  def diff(string1, string2) do
    string1
    |> String.myers_difference(string2)
    |> Enum.reduce(Delta.new(), fn
      {:eq, string}, delta -> Delta.retain(delta, Text.JS.length(string))
      {:ins, string}, delta -> Delta.insert(delta, string)
      {:del, string}, delta -> Delta.delete(delta, Text.JS.length(string))
    end)
    |> Delta.trim()
  end
end
